define source-git (#185)
packit/packit.dev: content/docs/source-git/_index.md

weight: 1
---

Let's talk about what source-git actually is.

We initiated source-git as an alternative to working in
[dist-git](https://github.com/release-engineering/dist-git). The need was
especially amplified after we realized that many projects already do this. The
problem was that everyone created a workflow and tooling for themselves which
made it really hard to share knowledge and onboard new projects. Our intention
is to create an implementation which is highly configurable, easy to start with

> and provdes both command-line tooling and a service which is tightly

Reviewer: ```suggestion
and provides both command-line tooling and a service which is tightly
```

> integration with GitHub and GitLab.

Reviewer: ```suggestion
integrated with GitHub and GitLab.
```

Source-git is a git repository which contains upstream sources and packaging

> files for a downstream distribtions, in our case Fedora Linux, CentOS Stream

Reviewer: ```suggestion
files for downstream distributions, in our case Fedora Linux, CentOS Stream
```

Author: no matter how many times I read it, I make so many stupid mistákos :D

> files for a downstream distribtions, in our case Fedora Linux, CentOS Stream

Reviewer: What do you mean by 'packaging files'? RPM spec file? A source-git repo usually contains also downstream patches.

Author: yes, precisely - okay, let me rephrase

and Red Hat Enterprise Linux.

The basic requirement on a source-git repo is that one needs to be able to

> create a SRPM out of a current checkout. We need several steps to achieve such

Reviewer: ```suggestion
create a SRPM out of a current checkout. We need several files to achieve such
```

Author: not sure about this one though it's true your suggestion is more concrete, mine is abstract

a thing:

* RPM Spec file

* Archive with sources

> * Spec file patches (if applicable)

Reviewer: Under "Spec file patch" I'd imagine a patch that modifies a spec file :-) I'd call it
- patches
- downstream patches
- patches to be applied in the spec file

(you can ignore this note if you don't like any of those or don't have any better)

Author: I like a combination of 2) and 3) :D

### Spec file

I'm going to answer the elephant in the room right away: "Do we really need to

> have a spec file in our upstream repo, or a source-git repo?!" — No, you don't.
> We just need to have a way to obtain the spec file, so:
> * You can download it from anywhere you want
> * You can generate it
> * A template spec file can be filled in during a build process
> * Storing the spec file directly in the repository is obviously the most convenient

Reviewer: ```suggestion
have a spec file in our upstream repo, or a source-git repo?!" — No, you don't, even storing the spec file directly in the repository is the most convenient.
We just need to have a way to obtain the spec file, so, you can:
* download it from anywhere you want
* generate it
* fill in a template spec file during a build process
```

* fill in a template spec file during a build process

### Archive

Packit needs to be able to get an archive for the current git checkout. By
default this is done using `git-archive` command and you can override this in
any way you want:
* Use autotools or any build system of your choice
* Use the native way of how archives are being generated in a particular ecosystem
* Script it yourself

> ### Spec file patches

Reviewer: Do we want to change this to "Downstream patches to be applied in the spec file" as well?
